### PR TITLE
Add SpeziLLMFog that performs dynamic LLM job dispatch to fog nodes

### PR DESCRIPTION
**Pull request by philippzagar**
# Add SpeziLLMFog that performs dynamic LLM job dispatch to fog nodes

## :recycle: Current situation & Problem
Currently, SpeziLLM doesn't support a privacy friendly execution of LLM inference jobs within the local network.

## :gear: Release Notes 
- Add SpeziLLMFog that performs dynamic LLM job dispatch to fog nodes.
- Include a FogNode component that easily starts up a complete fog node ready to perform LLM inference requests via docker compose.
- Introduce custom `LLMContext` type that holds the internal state of the `LLMSession`, instead of relying on the `SpeziChat` models.

## :books: Documentation
Added proper readmes, docs, and in-line comments

## :white_check_mark: Testing
Manual testing (for now)

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Comment by philippzagar** [comment]
@PSchmiedmayer The [Markdown link check](https://github.com/StanfordSpezi/SpeziLLM/actions/runs/8444431469/job/23129956955) is supposed to fail, as I'm already using SPM docs that are published as soon as this PR is merged.

**Comment by philippzagar** [comment]
@PSchmiedmayer Please adjust the merge checks of the repo as well, as I added additional testing jobs for iPad, Vision, and macOS

**Review by PSchmiedmayer — APPROVED** [review]
Thank you for all the work that went into this @philippzagar; greatly appreciated!

It looks great, happy to see this merged and completing the major LLM features we wanted to implement 🚀

File: Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunctionParameterWrapper.swift [hunk 1]
```diff
@@ -10,9 +10,9 @@ import OpenAI
 
 
 /// Alias of the OpenAI `JSONSchema/Property` type, describing properties within an object schema.
-public typealias LLMFunctionParameterPropertySchema = JSONSchema.Property
+public typealias LLMFunctionParameterPropertySchema = ChatQuery.ChatCompletionToolParam.FunctionDefinition.FunctionParameters.Property
 /// Alias of the OpenAI `JSONSchema/Item` type, describing array items within an array schema.
-public typealias LLMFunctionParameterItemSchema = JSONSchema.Items
+public typealias LLMFunctionParameterItemSchema = ChatQuery.ChatCompletionToolParam.FunctionDefinition.FunctionParameters.Property.Items
```
[the remark this answers is not in the document]
Comment (philippzagar) [reply Sources/SpeziLLMOpenAI/FunctionCalling/LLMFunctionParameterWrapper.swift:15]: Haha, let's see if we can use the OpenAPI generator at some point or another to get around this mess :D

**Comment by PSchmiedmayer** [comment]
Thank you @philippzagar for the PR and all the advancements! Thanks for adding all the additional automated tests! 🚀

**Pull request merged**